File: it.unibo.cautiousExplorer/app/src/main/java/it/unibo/executor/RunawayActor.java
package it.unibo.executor;

import it.unibo.interaction.IJavaActor;
import mapRoomKotlin.mapUtil;
import org.json.JSONObject;

import static it.unibo.executor.ApplMsgs.*;

/*
The map is a singleton object, managed by mapUtil
 */
public class RunawayActor extends PathExecutorActor {


    public RunawayActor(String name, IJavaActor ownerActor ) {
        super(name, ownerActor );
    }
    protected void updateTripInfo(String move){
        moves.updateMovesRep(move);
        mapUtil.doMove(move);
    }

    protected void resetStateVars(){
        curState          = State.start;
        moves.cleanMovesRepresentation();
        todoPath       = "";
    }
    protected void doMove(char moveStep){
        System.out.println("RunawayActor | doMove ... " + moveStep + " totPath="+todoPath);
        if( moveStep == 'w') doStep();
        else if( moveStep == 'l') turnLeft();
        else if( moveStep == 'r') turnRight();
        else if( moveStep == 's') doBackStep();
    }

    protected void fsmrunaway(String move, String endmove) {
        System.out.println(myname + " | state=" +
                curState +  " move=" + move + " endmove=" + endmove + " totPath="+todoPath );
        switch (curState) {
            case start: {
                if( move.equals(runawyStartId)) {
                    System.out.println("=&=&=&=&=RunawayActor runaway&=&=&=&=&=&=&=&=&=&=& ");
                }
                if (todoPath.length() > 0) {
                    mapUtil.showMap();
                    //waitUser();
                    doMove(todoPath.charAt(0));
                    curState = State.moving;
                } else curState = State.endok;
                break;
            }
            case moving: {
                String moveShort = MoveNameShort.get(move);

                if (endmove.equals("true")){
                    updateTripInfo(moveShort);
                    mapUtil.showMap();
                    todoPath = todoPath.substring(1);   //the first move has been done
                    System.out.println(myname + " | moveShort " + moveShort + " totPath=" + todoPath);
                    if( todoPath.length() > 0){
                        curState = State.moving;
                        doMove( todoPath.charAt(0) );
                    }else{ //totPath.length() == 0
                        microStep();
                        curState = State.endok;
                    }
                }else{  //endmove=false (obstacle)
                        System.out.println(myname + "|  FATAL ERROR: OUT OF HYPOTHESIS"  );

                        support.removeActor(this);
                        terminate();
                 }
                break;
            }//moving
            case endok: {
                System.out.println(myname + " | END OK ---------------- "  );
                ownerActor.send(runawyEndMsg);

                support.removeActor(this);
                terminate();
                //mapUtil.showMap();
                //resetStateVars();
                break;
            }//end


            default: {
                System.out.println(myname + " | error - curState = " + curState);
            }
        }

    }


/*
======================================================================================
 */
    @Override
    protected void msgDriven( JSONObject msgJson){
         if( msgJson.has(runawyStartId) ) {
             System.out.println("RunawayActor | runaway infoJson:" + msgJson);
             this.todoPath = msgJson.getString(runawyStartId);
             fsmrunaway( runawyStartId, "");
         }else if( msgJson.has(endMoveId) ) {
             System.out.println("RunawayActor | infoJson:" + msgJson);
             fsmrunaway(msgJson.getString("move"), msgJson.getString("endmove"));

         }
    }



}

File: it.unibo.kotlinSupports/app/src/main/kotlin/it/unibo/executors/RunawayActor.kt
package it.unibo.executors

 
import it.unibo.executors.ApplMsgs.endMoveId
import it.unibo.executors.ApplMsgs.runawyEndMsg
import it.unibo.executors.ApplMsgs.runawyStartId
import it.unibo.interaction.IJavaActor
import kotlinx.coroutines.CoroutineScope
import mapRoomKotlin.mapUtil
import org.json.JSONObject

/*
The map is a singleton object, managed by mapUtil
 */
class RunawayActor(name: String, ownerActor: IJavaActor, scope: CoroutineScope) : ExecutorActor(name, ownerActor, scope) {

    /*
    override fun updateTripInfo(move: String) {
        moves.updateMovesRep(move)
        mapUtil.doMove(move)
    }*/
    /*
    override fun resetStateVars() {
        curState = State.start
        moves.cleanMovesRepresentation()
        todoPath = ""
    }*/
    /*
    override fun doMove(moveStep: Char) {
        println("RunawayActor | doMove ... $moveStep totPath=$todoPath")
        if (moveStep == 'w') doStep() else if (moveStep == 'l') turnLeft() else if (moveStep == 'r') turnRight() else if (moveStep == 's') doBackStep()
    }*/

    protected fun fsmrunaway(move: String, endmove: String) {
        println(
            name + " | state=" +
                    curState + " move=" + move + " endmove=" + endmove + " totPath=" + todoPath
        )
        when (curState) {
            State.start -> {
                if (move == runawyStartId) {
                    println("=&=&=&=&=RunawayActor runaway&=&=&=&=&=&=&=&=&=&=& ")
                }
                if (todoPath.length > 0) {
                    mapUtil.showMap()
                    //waitUser();
                    doMove(todoPath.get(0))
                    curState = State.moving
                } else curState = State.endok
            }
            State.moving -> {
                val moveShort = MoveNameShort[move]
                if (endmove == "true") {
                    updateTripInfo(moveShort!!)
                    mapUtil.showMap()
                    todoPath = todoPath.substring(1) //the first move has been done
                    println("$name | moveShort $moveShort totPath=$todoPath")
                    if (todoPath.length > 0) {
                        curState = State.moving
                        doMove(todoPath.get(0))
                    } else { //totPath.length() == 0
                        microStep()
                        curState = State.endok
                    }
                } else {  //endmove=false (obstacle)
                    println("$name|  FATAL ERROR: OUT OF HYPOTHESIS")
                    support.removeActor(this)
                    terminate()
                }
            } //moving
            State.endok -> {
                println("$name | END OK ---------------- ")
                ownerActor.send(runawyEndMsg)
                support.removeActor(this)
                terminate()
            } //end
            else -> {
                println("$name | error - curState = $curState")
            }
        }
    }

    /*
======================================================================================
 */
    override fun msgDriven(msgJson: JSONObject) {
        if (msgJson.has(runawyStartId)) {
            println("RunawayActor | runaway infoJson:$msgJson")
            todoPath = msgJson.getString(runawyStartId)
            fsmrunaway(runawyStartId, "")
        } else if (msgJson.has(endMoveId)) {
            println("RunawayActor | infoJson:$msgJson")
            fsmrunaway(msgJson.getString("move"), msgJson.getString("endmove"))
        }
    }
}